### PR TITLE
#998 - Add vertices list for HPolyhedron

### LINKED ISSUE #998: Add vertices_list for HPolyhedron
Currently we throw an error. However, the implementation for `HPolytope` works for unbounded polytopes and does what you would expect.
I think we have decided to not support vertices of polyhedra, but since it is so easy, should we support it?
```julia
julia> using LazySets, Polyhedra

julia> constraints = [LazySets.HalfSpace([1., 0.], 0.), LazySets.HalfSpace([0., 1.], 0.)];

julia> vertices_list(HPolyhedron(constraints))
ERROR: ArgumentError: the list of vertices of a (possibly unbounded) polyhedron is not
 defined; if the polyhedron is bounded, try converting to `HPolytope` first
Stacktrace:
 [1] vertices_list(::HPolyhedron{Float64}) at LazySets/src/HPolyhedron.jl:631
 [2] top-level scope at none:0

julia> vertices_list(HPolytope(constraints))
1-element Array{Array{Float64,1},1}:
 [0.0, 0.0]
```

### PR DESCRIPTION
**Pull request by mforets**
Closes #998.

### CODE REVIEW

File: src/Interfaces/AbstractHyperrectangle.jl [hunk 1]
```diff
@@ -569,8 +569,8 @@ function split(H::AbstractHyperrectangle{N}, num_blocks::AbstractVector{Int}
     end
 
     # create hyperrectangles for every combination of the center points
-    result = Vector{Hyperrectangle{N}}(undef, total_number)
-    for (i, center) in enumerate(product(centers...))
+    result = Vector{Hyperrectangle{N, typeof(center(H)), typeof(radius)}}(undef, total_number)
```
Comment (mforets) [inline src/Interfaces/AbstractHyperrectangle.jl:572]: oops this was not part of the PR :)

File: src/Interfaces/AbstractPolyhedron_functions.jl [hunk 2]
```diff
@@ -879,3 +881,67 @@ function an_element(P::AbstractPolyhedron{N}) where {N<:Real}
...
+                         polyhedron is bounded
+
+### Output
+
+The list of vertices of `P`, or an error if `P` is unbounded.
+
+### Notes
+
+This function returns an error if the polyhedron is unbounded. Otherwise,
+the polyhedron is converted to an `HPolytpe` and its list of vertices is computed.
```
Comment (schillic) [inline src/Interfaces/AbstractPolyhedron_functions.jl:903]:
```suggestion
the polyhedron is converted to an `HPolytope` and its list of vertices is computed.
```

File: test/unit_Polyhedron.jl [hunk 3]
```diff
@@ -75,7 +75,7 @@ for N in [Float64, Rational{Int}, Float32]
         # convert hyperrectangle to a HPolyhedron
         H = Hyperrectangle(N[1, 1], N[2, 2])
         P = convert(HPolyhedron, H)
-        @test_throws ArgumentError vertices_list(P) # the vertices list is not defined, see #820
+        @test vertices_list(P) == 4
```
Comment (mforets) [inline test/unit_Polyhedron.jl:78]:
```suggestion
        @test length(vertices_list(P)) == 4
```

File: docs/src/lib/interfaces.md [hunk 4]
```diff
@@ -177,6 +177,8 @@ constrained_dimensions(::AbstractPolyhedron)
 linear_map(::AbstractMatrix{N}, ::AbstractPolyhedron{N}) where {N<:Real}
 chebyshev_center(::AbstractPolyhedron{N}) where {N<:AbstractFloat}
 an_element(::AbstractPolyhedron{N}) where {N<:Real}
+vertices_list(::AbstractPolyhedron, ::Bool=true)
+singleton_list(::AbstractPolyhedron, ::Bool=false)
```
Comment (mforets) [inline docs/src/lib/interfaces.md:181]:
```suggestion
singleton_list(::AbstractPolyhedron, ::Bool=true)
```

**Pull request merged**